File: Java/Reverse-Linked-List-II.java
/**
 * Definition for ListNode
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    /**
     * @param ListNode head is the head of the linked list 
     * @oaram m and n
     * @return: The head of the reversed ListNode
     */
    public ListNode reverseBetween(ListNode head, int m , int n) {
        // write your code
        if (m == n) return head;
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode start = dummy;
        for (int i = 0; i < m - 1; i ++) {
        	start = start.next;
        }
        ListNode cur = start.next;
        ListNode end = cur;
        for (int i = 0; i < n - m; i ++) {
        	end = end.next;
        }
        ListNode cnt = end.next;
        end.next = null;
        while (cur != null) {
        	ListNode temp = cur.next;
        	cur.next = cnt;
        	cnt = cur;
        	cur = temp;
        }
        start.next = cnt;
        return dummy.next;
    }
}

File: Cpp/Reverse-Linked-List-II.cpp
#include <stack>

/**
 * Definition of singly-linked-list:
 */
class ListNode {
public:
    int val;
    ListNode *next;
    ListNode(int val) {
        this->val = val;
        this->next = NULL;
    }
};

class Solution {
public:
    /**
     * @param head: The head of linked list.
     * @param m: The start position need to reverse.
     * @param n: The end position need to reverse.
     * @return: The new head of partial reversed linked list.
     */
    ListNode *reverseBetween(ListNode *head, int m, int n) {
        // write your code here
        if (m == n) return head;
        ListNode *cnt = head;
        ListNode *pre = new ListNode(-1), *res = pre;
        for (int i = 0; i < m - 1; i ++) {
            res->next = cnt;
            res = cnt;
            cnt = cnt->next;
        }
        stack<ListNode*> st;
        for (int i = 0; i <= n - m; i ++) {
            st.push(cnt);
            cnt = cnt->next;
        }
        while (!st.empty()) {
            res->next = st.top();
            res = res->next;
            st.pop();
        }
        res->next = cnt;
        return pre->next;
    }
};

